pkg/operator/clustermemberremovalcontroller/clustermemberremovalcontroller.go

```diff
@@ -0,0 +1,227 @@
+func NewClusterMemberRemovalController(
+	isMachineAPIFunctional func() (bool, error),
+	etcdClient etcdcli.EtcdClient,
+	masterNodeInformer cache.SharedInformer,
+	masterMachineInformer cache.SharedInformer,
+	masterNodeLister corev1listers.NodeLister,
+	masterMachineLister machinelistersv1beta1.MachineLister,
+	hasNetworkInformerSyncedFn func() bool,
+	networkLister configv1listers.NetworkLister,
+	eventRecorder events.Recorder,
+) factory.Controller {
+	c := &ClusterMemberRemovalController{
+		etcdClient:                 etcdClient,
+		isMachineAPIFunctionalFn:   isMachineAPIFunctional,
+		masterNodeLister:           masterNodeLister,
+		masterMachineLister:        masterMachineLister,
+		networkLister:              networkLister,
+		hasNetworkInformerSyncedFn: hasNetworkInformerSyncedFn,
+	}
+	return factory.New().
+		WithSync(c.sync).
+		ResyncEvery(33*time.Minute). // make it slow since nodes are updated every few minutes
+		WithInformers(
+			masterNodeInformer,
+			masterMachineInformer,
+		).ToController("ClusterMemberRemovalController", eventRecorder.WithComponentSuffix("cluster-member-removal-controller"))
+}
+
+func (c *ClusterMemberRemovalController) sync(ctx context.Context, _ factory.SyncContext) error {
+	if !c.hasNetworkInformerSyncedFn() {
+		return nil
+	}
+	// stop if the machine API is not functional
+	if isFunctional, err := c.isMachineAPIFunctionalFn(); err != nil {
+		return err
+	} else if !isFunctional {
+		return nil
+	}
+	// we are only interested in unhealthy members
+	// because theirs nodes are gone
+	unhealthyMembers, err := c.etcdClient.UnhealthyMembers(ctx)
+	if err != nil {
+		return err
+	}
+	if len(unhealthyMembers) == 0 {
+		return nil
+	}
+	for _, unhealthyMember := range unhealthyMembers {
+		memberURL, err := memberToURL(unhealthyMember)
+		if err != nil {
+			klog.Errorf("%v", err)
+			continue
+		}
+		memberLocator := fmt.Sprintf("%v/%v/%v", memberURL, unhealthyMember.Name, unhealthyMember.ID)
+
+		machine, err := c.getMachineForMember(unhealthyMember)
+		if err != nil {
+			klog.Errorf("unable to get a machine for the unhealthy member: %v, err: %v", memberLocator, err)
+			continue
+		}
+		if machine != nil {
+			klog.V(4).Infof("cannot remove the unhealthy member: %v because a machine resource still exists (%v/%v)", memberLocator, machine.Name, machine.UID)
+			continue
+		}
+
+		node, err := c.getNodeForMember(unhealthyMember)
+		if err != nil {
+			klog.Errorf("unable to get a node for the unhealthy member: %v, err: %v", memberLocator, err)
+			continue
+		}
+		if node != nil {
+			klog.V(4).Infof("cannot remove the unhealthy member: %v because a node resource still exists (%v/%v)", memberLocator, node.Name, node.UID)
+			continue
+		}
+
+		if len(unhealthyMember.Name) == 0 {
+			klog.Errorf("unable to remove the unhealthy member: %v because it doesn't have a name (has the member been started?)", memberLocator)
```

Comment:
What if it never started but the Machine has now been deleted? Can we end up in a lock here?

Reply:
good point, I don't have any experience with running etcd. @hasbro17 could you please share your thoughts on that? Is this something we need to take into account? Is there a way of removing a member with an empty Name field?